component/src/main/java/org/wso2/extension/siddhi/execution/streamingml/clustering/kmeans/util/Clusterer.java

```diff
@@ -0,0 +1,346 @@
+                assignToCluster(dataPointsArray);
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Current model : \n" + model.getModelInfo());
+                }
+                List<Cluster> newClusterList = calculateNewClusters();
+
+                centroidShifted = !model.getClusterList().equals(newClusterList);
+                if (logger.isDebugEnabled()) {
+                    logger.debug("previous model : " + printClusterList(model.getClusterList()) + "\nnew model : " +
+                            printClusterList(newClusterList) + "\ncentroid shifted?" + centroidShifted);
+                }
+                if (!centroidShifted) {
+                    break;
+                }
+                model.setClusterList(newClusterList);
+                iter++;
+            }
+        }
+    }
+
+    public String printClusterList(List<Cluster> clusterList) {
+        StringBuilder s = new StringBuilder();
+        for (Cluster c: clusterList) {
+            s.append(Arrays.toString(c.getCentroid().getCoordinates()));
+        }
+        return s.toString();
+    }
+
+    public void buildModel(List<DataPoint> dataPointsArray) {
+        int distinctCount = model.size();
+        for (DataPoint currentDataPoint : dataPointsArray) {
+            if (distinctCount >= numberOfClusters) {
+                break;
+            }
+            DataPoint coordinatesOfCurrentDataPoint = new DataPoint();
+            coordinatesOfCurrentDataPoint.setCoordinates(currentDataPoint.getCoordinates());
+            if (!model.contains(coordinatesOfCurrentDataPoint)) {
+                model.add(coordinatesOfCurrentDataPoint);
+                distinctCount++;
+            }
+        }
+    }
+
+    /**
+     * After the first clustering this method can be used to incrementally update centroidList
+     * in real time. This method takes in the new set of datapoints and decayRate as inputs
+     * and calculates the centroids of the new set. Then new centroids are calculated using
+     * newAvg = oldAvg + decayRate * batchAvg
+     *
+     * @param dataPointsArray
+     * @param decayRate       should be in [0,1]
+     */
+    public void updateCluster(List<DataPoint> dataPointsArray, double decayRate) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Updating cluster");
+            logger.debug("model at the start of this update : ");
+            logger.debug(model.getModelInfo());
+        }
+        StringBuilder s;
+        List<Cluster> intermediateClusterList = new LinkedList<>();
+
+        int iter = 0;
+        if (dataPointsArray.size() != 0) {
+            //when number of elements in centroid list is less than numberOfClusters
+            if (model.size() < numberOfClusters) {
+                buildModel(dataPointsArray);
+            }
+            if (model.size() == numberOfClusters) {
+                ArrayList<Cluster> oldClusterList = new ArrayList<>(numberOfClusters);
```

Comment:
Make it   List<Cluster> oldClusterList = new ArrayList<>(numberOfClusters)

Reply:
Fixed in https://github.com/siddhi-io/siddhi-execution-streamingml/pull/76